File: android/src/main/java/com/mobilecoin/mobilecoin_flutter/FfiPrintableWrapper.java
package com.mobilecoin.mobilecoin_flutter;

import com.mobilecoin.lib.PaymentRequest;
import com.mobilecoin.lib.PrintableWrapper;
import com.mobilecoin.lib.TransferPayload;
import com.mobilecoin.lib.PublicAddress;
import com.mobilecoin.lib.exceptions.SerializationException;

import androidx.annotation.Keep;

import java.util.Objects;

@Keep
public class FfiPrintableWrapper {

    private FfiPrintableWrapper() {
    }

    public static int fromB58String(String b58String) throws SerializationException {
        PrintableWrapper printableWrapper = PrintableWrapper.fromB58String(b58String);
        final int hashCode = printableWrapper.hashCode();
        ObjectStorage.addObject(hashCode, printableWrapper);
        return hashCode;
    }

    public static String toB58String(int wrapperId) throws SerializationException {
        PrintableWrapper printableWrapper = (PrintableWrapper) ObjectStorage.objectForKey(wrapperId);
        return printableWrapper.toB58String();
    }

    public static int fromPublicAddress(int publicAddressId) throws SerializationException {
        PublicAddress publicAddress = (PublicAddress) ObjectStorage.objectForKey(publicAddressId);
        PrintableWrapper printableWrapper = PrintableWrapper.fromPublicAddress(publicAddress);
        final int hashCode = printableWrapper.hashCode();
        ObjectStorage.addObject(hashCode, printableWrapper);
        return hashCode;
    }

    public static boolean hasPublicAddress(int wrapperId) {
        PrintableWrapper printableWrapper = (PrintableWrapper) ObjectStorage.objectForKey(wrapperId);
        return printableWrapper.hasPublicAddress();
    }

    public static int getPublicAddress(int wrapperId) {
        PrintableWrapper printableWrapper = (PrintableWrapper) ObjectStorage.objectForKey(wrapperId);
        PublicAddress publicAddress = printableWrapper.getPublicAddress();
        Objects.requireNonNull(publicAddress);
        final int hashCode = publicAddress.hashCode();
        ObjectStorage.addObject(hashCode, publicAddress);
        return hashCode;
    }

    public static int fromTransferPayload(int payloadId) throws SerializationException {
        TransferPayload transferPayload = (TransferPayload) ObjectStorage.objectForKey(payloadId);
        PrintableWrapper printableWrapper = PrintableWrapper.fromTransferPayload(transferPayload);
        final int hashCode = printableWrapper.hashCode();
        ObjectStorage.addObject(hashCode, printableWrapper);
        return hashCode;
    }

    public static boolean hasTransferPayload(int wrapperId) {
        PrintableWrapper printableWrapper = (PrintableWrapper) ObjectStorage.objectForKey(wrapperId);
        return printableWrapper.hasTransferPayload();
    }

    public static int getTransferPayload(int wrapperId) {
        PrintableWrapper printableWrapper = (PrintableWrapper) ObjectStorage.objectForKey(wrapperId);
        TransferPayload transferPayload = printableWrapper.getTransferPayload();
        Objects.requireNonNull(transferPayload);
        final int hashCode = transferPayload.hashCode();
        ObjectStorage.addObject(hashCode, transferPayload);
        return hashCode;
    }

    public static int fromPaymentRequest(int paymentRequestId) throws SerializationException {
        PaymentRequest paymentRequest = (PaymentRequest) ObjectStorage.objectForKey(paymentRequestId);
        PrintableWrapper printableWrapper = PrintableWrapper.fromPaymentRequest(paymentRequest);
        final int hashCode = printableWrapper.hashCode();
        ObjectStorage.addObject(hashCode, printableWrapper);
        return hashCode;
    }
    public static boolean hasPaymentRequest(int wrapperId) {
        PrintableWrapper printableWrapper = (PrintableWrapper) ObjectStorage.objectForKey(wrapperId);
        return printableWrapper.hasPaymentRequest();
    }

    public static int getPaymentRequest(int wrapperId) {
        PrintableWrapper printableWrapper = (PrintableWrapper) ObjectStorage.objectForKey(wrapperId);
        PaymentRequest paymentRequest = printableWrapper.getPaymentRequest();
        Objects.requireNonNull(paymentRequest);
        final int hashCode = paymentRequest.hashCode();
        ObjectStorage.addObject(hashCode, paymentRequest);
        return hashCode;
    }
}

File: ios/mobilecoin_flutter/Sources/mobilecoin_flutter/FfiPrintableWrapper.swift
//
//  FfiPrintableWrapper.swift
//  mobilecoin_flutter
//
//  Created by Alexander Voloshyn on 9/29/20.
//

import Flutter
import Foundation
import MobileCoin

struct FfiPrintableWrapper{
    struct FromB58String: Command {
        func execute(args: [String : Any], result: @escaping FlutterResult) throws {
            guard let b58String: String = args["b58String"] as? String else {
                throw PluginError.invalidArguments
            }
            let decoded = Base58Coder.decode(b58String)
            switch decoded {
            case .publicAddress(let publicAddress):
                let hashCode: Int = publicAddress.hashValue
                ObjectStorage.addObject(publicAddress, forKey: hashCode)
                result(hashCode)
                break;
            case .transferPayload(let transferPayload):
                let hashCode: Int = transferPayload.hashValue
                ObjectStorage.addObject(transferPayload, forKey: hashCode)
                result(hashCode)
                break;
            case .paymentRequest(let paymentRequest):
                let hashCode: Int = paymentRequest.hashValue
                ObjectStorage.addObject(paymentRequest, forKey: hashCode)
                result(hashCode)
                break;
            default:
                throw PluginError.unimplemented
            }
        }
    }
    struct ToB58String: Command {
        func execute(args: [String : Any], result: @escaping FlutterResult) throws {
            guard let id: Int = args["id"] as? Int,
                  let obj = ObjectStorage.objectForKey(id) else {
                      throw PluginError.invalidArguments
                  }
            if let publicAddress = obj as? PublicAddress {
                let b58String = Base58Coder.encode(publicAddress)
                result(b58String)
            } else if let transferPayload = obj as? TransferPayload {
                let b58String = Base58Coder.encode(transferPayload)
                result(b58String)
            } else if let paymentRequest = obj as? PaymentRequest {
                let b58String = Base58Coder.encode(paymentRequest)
                result(b58String)
            } else {
                throw PluginError.unimplemented
            }
        }
    }
    struct FromPublicAddress: Command {
        func execute(args: [String : Any], result: @escaping FlutterResult) throws {
            if let id: Int = args["id"] as? Int {
                result(id)
            } else {
                throw PluginError.invalidArguments
            }
            
        }
    }
    struct HasPublicAddress: Command {
        func execute(args: [String : Any], result: @escaping FlutterResult) throws {
            guard let id: Int = args["id"] as? Int else {
                throw PluginError.invalidArguments
            }
            if let _ = ObjectStorage.objectForKey(id) as? PublicAddress {
                result(true)
            } else {
                result(false)
            }
        }
    }
    struct GetPublicAddress: Command {
        func execute(args: [String : Any], result: @escaping FlutterResult) throws {
            guard let id: Int = args["id"] as? Int else {
                throw PluginError.invalidArguments
            }
            if let _ = ObjectStorage.objectForKey(id) as? PublicAddress {
                result(id)
            } else {
                throw PluginError.notFound
            }
        }
    }
    struct FromTransferPayload: Command {
        func execute(args: [String : Any], result: @escaping FlutterResult) throws {
            guard let id: Int = args["id"] as? Int else {
                throw PluginError.invalidArguments
            }
            result(id)
        }
    }
    struct HasTransferPayload: Command {
        func execute(args: [String : Any], result: @escaping FlutterResult) throws {
            guard let id: Int = args["id"] as? Int else {
                throw PluginError.invalidArguments
            }
            if let _ = ObjectStorage.objectForKey(id) as? TransferPayload {
                result(true)
            } else {
                result(false)
            }
        }
    }
    struct GetTransferPayload: Command {
        func execute(args: [String : Any], result: @escaping FlutterResult) throws {
            guard let id: Int = args["id"] as? Int else {
                throw PluginError.invalidArguments
            }
            if let _ = ObjectStorage.objectForKey(id) as? TransferPayload {
                result(id)
            } else {
                throw PluginError.notFound
            }
        }
    }
    struct FromPaymentRequest: Command {
        func execute(args: [String : Any], result: @escaping FlutterResult) throws {
            guard let id: Int = args["id"] as? Int else {
                throw PluginError.invalidArguments
            }
            result(id)
        }
    }
    struct HasPaymentRequest: Command {
        func execute(args: [String : Any], result: @escaping FlutterResult) throws {
            guard let id: Int = args["id"] as? Int else {
                throw PluginError.invalidArguments
            }
            if let _ = ObjectStorage.objectForKey(id) as? PaymentRequest {
                result(true)
            } else {
                result(false)
            }
        }
    }
    struct GetPaymentRequest: Command {
        func execute(args: [String : Any], result: @escaping FlutterResult) throws {
            guard let id: Int = args["id"] as? Int else {
                throw PluginError.invalidArguments
            }
            if let _ = ObjectStorage.objectForKey(id) as? PaymentRequest {
                result(id)
            } else {
                throw PluginError.notFound
            }
        }
    }
}


